solidity/contracts/bank/Bank.sol

```diff
@@ -0,0 +1,299 @@
+            _increaseBalance(recipients[i], amounts[i]);
+        }
+    }
+
+    /// @notice Increases balance of the provided `recipient` by the provided
+    ///         `amount`. Can only be called by the Bridge.
+    function increaseBalance(address recipient, uint256 amount)
+        external
+        onlyBridge
+    {
+        _increaseBalance(recipient, amount);
+    }
+
+    /// @notice Decreases caller's balance by the provided `amount`. There is no
+    ///         way to restore the balance so do not call this function unless
+    ///         you really know what you are doing!
+    function decreaseBalance(uint256 amount) external {
+        balanceOf[msg.sender] -= amount;
+        emit BalanceDecreased(msg.sender, amount);
+    }
+
+    /// @notice Returns hash of EIP712 Domain struct with `TBTC Bank` as
+    ///         a signing domain and Bank contract as a verifying contract.
+    ///         Used to construct EIP2612 signature provided to `permit`
+    ///         function.
+    /* solhint-disable-next-line func-name-mixedcase */
+    function DOMAIN_SEPARATOR() public view returns (bytes32) {
+        // As explained in EIP-2612, if the DOMAIN_SEPARATOR contains the
+        // chainId and is defined at contract deployment instead of
+        // reconstructed for every signature, there is a risk of possible replay
+        // attacks between chains in the event of a future chain split.
+        // To address this issue, we check the cached chain ID against the
+        // current one and in case they are different, we build domain separator
+        // from scratch.
+        if (block.chainid == cachedChainId) {
+            return cachedDomainSeparator;
+        } else {
+            return buildDomainSeparator();
+        }
+    }
+
+    function _increaseBalance(address recipient, uint256 amount) internal {
+        require(
+            recipient != address(this),
+            "Can not increase balance for Bank"
+        );
+        balanceOf[recipient] += amount;
+        emit BalanceIncreased(recipient, amount);
+    }
+
+    function _transferBalance(
+        address spender,
+        address recipient,
+        uint256 amount
+    ) private {
+        require(
+            recipient != address(0),
+            "Can not transfer to the zero address"
+        );
+        require(
+            recipient != address(this),
+            "Can not transfer to the Bank address"
+        );
+
+        uint256 spenderBalance = balanceOf[spender];
+        require(spenderBalance >= amount, "Transfer amount exceeds balance");
+        balanceOf[spender] = spenderBalance - amount;
+        balanceOf[recipient] += amount;
+        emit BalanceTransferred(spender, recipient, amount);
+    }
+
+    function _approveBalance(
+        address owner,
+        address spender,
+        uint256 amount
+    ) private {
+        require(spender != address(0), "Can not approve to the zero address");
```

Comment:
I see we prevent against transferring to the bank address and increasing their balance. Shouldn't we protect against allowances having the bank address as spender for consistency?

Reply:
We can but I think it does not hurt that much as transferring to the Bank. If one transfers their balance to the Bank, it is unrecoverable. If they approve Bank to use their balances, they will only lose gas. Bank can not be the `msg.sender` for `Bank.transferBalanceFrom` so I *think* there is no security risk.

Reply:
Ok sure, let's leave as is.